combined_robot_hw_tests/CMakeLists.txt

```diff
@@ -36,6 +36,7 @@ if(CATKIN_ENABLE_TESTING)
 
   add_executable(combined_robot_hw_dummy_app EXCLUDE_FROM_ALL src/dummy_app.cpp)
   target_link_libraries(combined_robot_hw_dummy_app ${PROJECT_NAME} ${catkin_LIBRARIES})
+  add_dependencies(tests combined_robot_hw_dummy_app)
```

Comment:
I guess this should better be wrapped with `if(TARGET tests)` as in `ros_comm`.

Reply:
I'd rather move it into the `if(CATKIN_ENABLE_TESTING)` section, I strongly dislike generating more cmake blurb

Reply:
> I'd rather move it into the if(CATKIN_ENABLE_TESTING)

It already is.

> I strongly dislike generating more cmake blurb

I don't understand why we moved to `ros_comm` style in the first place.
A simpler alternative is to just remove `EXCLUDE_FROM_ALL`.
The packages in question a test packages anyway.

Reply:
>A simpler alternative is to just remove EXCLUDE_FROM_ALL.
The packages in question are test packages anyway.

True, but I think the extra safeguards against installing test targets is worthwhile.

So, the conclusion here is to add the `if(TARGET tests)`?

Reply:
A final note: `CATKIN_ENABLE_TESTING` defaults to `ON`, so the `EXCLUDE_FROM_ALL` is quite useful for speeding up regular builds and only building tests when appropriate.

---

Back to the `if(TARGET tests)`:

> So, the conclusion here is to add the if(TARGET tests)?

I dug into this more and changed my mind, I don't think we should add this.

I don't know why `ros_comm` uses it around their `add_dependencies(tests ...)`. The Catkin CMake library will _always_ create the `tests` target ([tests.cmake L81](https://github.com/ros/catkin/blob/b6410a5257bb991481b28570899830704d768979/cmake/test/tests.cmake#L81)) unless `CATKIN_ENABLE_TESTING` is set to `OFF`. That same file also implies that it is sufficient to use only the `CATKIN_ENABLE_TESTING` check for test stuff ([tests.cmake L34](https://github.com/ros/catkin/blob/b6410a5257bb991481b28570899830704d768979/cmake/test/tests.cmake#L34)).

Additionally, the `catkin_add_gtest()` and `_gmock()` functions don't do this check before they `add_dependencies(tests ...)`; They only check for `CATKIN_ENABLE_TESTING`, and issue a warning if they are called while `CATKIN_ENABLE_TESTING` is `OFF`.

Since we are already within `if(CATKIN_ENABLE_TESTING)`, I think we are safe without `if(TARGET tests)`. If we want more clarity on why `ros_comm` uses it, we could just ask the `ros_comm` maintainers why they use the check and make our decision accordingly.

Reply:
In fact, it looks like `ros_comm` even uses `if(TARGET ...)` for targets they just created the line before... See [test_rosbag L37](https://github.com/ros/ros_comm/blob/902fb00fc7b2e881575a270a99c3a077ba5cdbba/test/test_rosbag/CMakeLists.txt#L37). So their usage isn't specific to the `tests` target.

Reply:
> True, but I think the extra safeguards against installing test targets is worthwhile.

We are talking about `*_test`  packages, their only purpose is to provide everything for the tests.

> Additionally, the catkin_add_gtest() and _gmock() functions don't do this check before they add_dependencies(tests ...)

[They do it for executables](https://github.com/ros/catkin/blob/dfc93c72d9799f57442d08c2f6e9fd494bba95c1/cmake/test/gtest.cmake#L82)

Reply:
> In fact, it looks like ros_comm even uses if(TARGET ...) for targets they just created the line before

Yes, I guess it is best practice to check if the *external* target exists before using it.

Reply:
> We are talking about *_test packages, their only purpose is to provide everything for the tests.

Indeed, but all the better if they can do so while installing as few artifacts as possible (ie. only plugins, which are required to be installed by pluginlib).

Also, while we may disagree on this point, I think the build speedup and more expected behaviour (only building these targets in test builds) still makes this worthwhile.

> [They do it for executables](https://github.com/ros/catkin/blob/dfc93c72d9799f57442d08c2f6e9fd494bba95c1/cmake/test/gtest.cmake#L82)

This line is checking that the newly created target exists (Similar to [test_rosbag L37](https://github.com/ros/ros_comm/blob/902fb00fc7b2e881575a270a99c3a077ba5cdbba/test/test_rosbag/CMakeLists.txt#L37)), not checking that the `tests` target exists, as we're debating.